### PR TITLE
Configuring iptables firewall

### LINKED ISSUE #5: Configuring iptables firewall
Configuring iptables firewall for 80 and 443

### PR DESCRIPTION
**Pull request by masoodkamyab**

### CODE REVIEW

File: do.sh [hunk 1]
```diff
@@ -4,7 +4,11 @@
 echo "Starting system preparation..."
 apt update
 apt -y upgrade
-apt -y install vim
+echo "iptables-persistent iptables-persistent/autosave_v4 boolean true" \
+  > debconf-set-selections
+echo "iptables-persistent iptables-persistent/autosave_v6 boolean true" \
+  > debconf-set-selections
+apt -y install vim iptables-persistent
```
Comment (pylover) [inline do.sh]: One item per line:

```bash
apt install -y \
  vim \
  foo \
  bar
```

File: do.sh [hunk 2]
```diff
@@ -17,4 +21,17 @@ if ! grep -q "set editing-mode vi" "${HOME}/.inputrc" 2>/dev/null; then
 fi
 
 
+echo "Configuring iptables firewall..."
```
Comment (pylover) [inline do.sh]: I just wanted to let you know that I need more attention. iptables are dangerous. , you can ban yourself forever with an invalid iptables rule.

File: do.sh [hunk 3]
```diff
@@ -17,4 +21,17 @@ if ! grep -q "set editing-mode vi" "${HOME}/.inputrc" 2>/dev/null; then
...
+iptables -X
+iptables -P INPUT DROP
+iptables -P FORWARD DROP
+iptables -P OUTPUT ACCEPT
+iptables -A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+iptables -A INPUT -i lo -j ACCEPT
+iptables -A INPUT -p tcp --dport "${SSH_PORT}" -j ACCEPT
+iptables -A INPUT -p tcp --dport 80 -j ACCEPT
+iptables -A INPUT -p tcp --dport 443 -j ACCEPT
+netfilter-persistent save
```
Comment (pylover) [inline do.sh]: WTF!

Use `iptables-save > /etc/iptables/rules.v4`

File: do.sh [hunk 4]
```diff
@@ -17,4 +21,17 @@ if ! grep -q "set editing-mode vi" "${HOME}/.inputrc" 2>/dev/null; then
 fi
 
 
+echo "Configuring iptables firewall..."
+iptables -F
+iptables -X
+iptables -P INPUT DROP
```
Comment (pylover) [inline do.sh]: Use a rollback(async) timer before setting the `DROP` policy. and change it to `ACCEP` if the timer is not stopped with the user's confirmation.

File: do.sh [hunk 5]
```diff
@@ -17,4 +21,17 @@ if ! grep -q "set editing-mode vi" "${HOME}/.inputrc" 2>/dev/null; then
...
+iptables -P FORWARD DROP
+iptables -P OUTPUT ACCEPT
+iptables -A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
+iptables -A INPUT -i lo -j ACCEPT
```
Comment (pylover) [inline do.sh]: So relax!, more limitation please.

```iptables
-A INPUT -s 127.0.0.0/8 -j ACCEPT
```

File: do.sh [hunk 6]
```diff
@@ -17,4 +21,17 @@ if ! grep -q "set editing-mode vi" "${HOME}/.inputrc" 2>/dev/null; then
...
+iptables -A INPUT -p tcp --dport "${SSH_PORT}" -j ACCEPT
+iptables -A INPUT -p tcp --dport 80 -j ACCEPT
```
Comment (pylover) [inline do.sh]: Limit the source port too: `--sport 1024:65535`.

File: do.sh [hunk 7]
```diff
@@ -17,4 +21,17 @@ if ! grep -q "set editing-mode vi" "${HOME}/.inputrc" 2>/dev/null; then
 fi
 
 
+echo "Configuring iptables firewall..."
+iptables -F
+iptables -X
+iptables -P INPUT DROP
+iptables -P FORWARD DROP
+iptables -P OUTPUT ACCEPT
+iptables -A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
```
Comment (pylover) [inline do.sh]: So weird!

Use these:

```iptables
-A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
```

File: do.sh [hunk 8]
```diff
@@ -17,4 +21,17 @@ if ! grep -q "set editing-mode vi" "${HOME}/.inputrc" 2>/dev/null; then
...
+iptables -A INPUT -i lo -j ACCEPT
+iptables -A INPUT -p tcp --dport "${SSH_PORT}" -j ACCEPT
```
Comment (pylover) [inline do.sh]:
```iptables
-A INPUT -d x.x.x.x/32 -p tcp -m tcp --sport 1025:65535 --dport 80 -j ACCEPT
```

**Review by pylover — CHANGES_REQUESTED** [review]

File: do.sh (hunk 8 above)
Comment (masoodkazemi94) [reply do.sh]: Done

File: do.sh (hunk 7 above)
Comment (masoodkazemi94) [reply do.sh]: Done

File: do.sh (hunk 6 above)
Comment (masoodkazemi94) [reply do.sh]: Done

File: do.sh (hunk 5 above)
Comment (masoodkazemi94) [reply do.sh]: Done

File: do.sh (hunk 4 above)
Comment (masoodkazemi94) [reply do.sh]: Done

File: do.sh (hunk 3 above)
Comment (masoodkazemi94) [reply do.sh]: Done

File: do.sh (hunk 2 above)
Comment (masoodkazemi94) [reply do.sh]: Sure

File: do.sh (hunk 1 above)
Comment (masoodkazemi94) [reply do.sh]: Done

File: do.sh [hunk 9]
```diff
@@ -17,4 +28,72 @@ if ! grep -q "set editing-mode vi" "${HOME}/.inputrc" 2>/dev/null; then
...
 
+# Back up current iptables configuration
+echo "Backing up current iptables rules..."
+mkdir -p /etc/iptables/
+iptables-save > /etc/iptables/rules.v4.backup
+
+
+# Rollback script file
+ROLLBACK_FILE="/tmp/firewall_rollback.sh"
+cat <<EOT > "${ROLLBACK_FILE}"
```
Comment (pylover) [inline do.sh]: Ues echo |

File: do.sh [hunk 10]
```diff
@@ -17,4 +28,72 @@ if ! grep -q "set editing-mode vi" "${HOME}/.inputrc" 2>/dev/null; then
...
+
+
+# Start a screen session to ensure rollback can be applied if disconnected.
+# If no confirmation is provided by the user (or user disconnects), after $ROLLBACK_TIMER seconds rollback is triggered.
+screen -dmS ${SCREEN_NAME} bash -c "sleep ${ROLLBACK_TIMER} && bash ${ROLLBACK_FILE} && echo 'Firewall rollback applied due to timeout.'"
+
+
+# Apply new iptables rules
+echo "Applying new iptables firewall rules..."
+iptables -F
```
Comment (pylover) [inline do.sh]: Do not flush.

**Review by pylover — CHANGES_REQUESTED** [review]

File: do.sh (hunk 10 above)
Comment (masoodkazemi94) [reply do.sh]: Done.

**Pull request merged**